src/uu/env/src/raw_string_parser.rs

```diff
@@ -0,0 +1,307 @@
+// This file is part of the uutils coreutils package.
+//
+// For the full copyright and license information, please view the LICENSE
+// file that was distributed with this source code.
+//
+//! SAFETY: This module does "unsafe" byte by byte operations on a UTF8 encoded string.
```

Comment:
This should not be necessary. You should probably write it so that it operates not on `&str`, but on `&[u8]`, which should remove all the `unsafe` code. This matches what GNU does better (GNU code almost never requires valid UTF-8) and will probably be easier. I think you might even be able to remove most of this module.

Reply:
As a sidenote, I do like that you've split the implementation in safe and unsafe! That's a neat way of organizing this.

Reply:
Thanks for reviewing and feedback.
I'm happy to learn new things.
But before I start with the refactoring, I would like to know your opinion on these points:

1. As I understood, the input (program arguments) and the output (environmental variables and program name + arguments) is provided and required as `OsString`/`OsStr`. But anyway, I found in the existing code base of env a lot of cases where people are just using `str` or `String` to represent these values. E.g.

original version of env.rs - line 28 (all `str` and not `OsStr`):
```rust
struct Options<'a> {
    ignore_env: bool,
    line_ending: LineEnding,
    running_directory: Option<&'a str>,
    files: Vec<&'a str>,
    unsets: Vec<&'a str>,
    sets: Vec<(&'a str, &'a str)>,
    program: Vec<&'a str>,
}
```

or line 181 ( consider the `get_many::<String>`:
```rust
let running_directory = matches.get_one::<String>("chdir").map(|s| s.as_str());
    let files = match matches.get_many::<String>("file") {
        Some(v) => v.map(|s| s.as_str()).collect(),
        None => Vec::with_capacity(0),
    };
    let unsets = match matches.get_many::<String>("unset") {
        Some(v) => v.map(|s| s.as_str()).collect(),
        None => Vec::with_capacity(0),
    };
```

As we are adviced to be GNU compatible, we should stay away from `str` and `String` completely and purely use OsStr/OString. Is this the way to go?

Because what I also saw in other shell syntax parsing crates is that `str.to_bytes()` and `str::from_utf8_unchecked(bytes)` is used to convert back and forth string to bytes and bytes to string arguing that if it was valid before, it should also be valid afterwards. AND: if it was invalid before, it remains invalid exactly like before.

Thats why I invented this class. I tried to make this a bit more systematic.
Additionally, its not quite convenient to debug with byte-vectors. Because the debugger doesn't interpret the values as characters but as numbers instead. This way a developer needs to manually convert numbers to chars to understand if the code works fine or not. Do you have a solution for this?

Reply:
Yes everythinng should be`OsStr`, but there a lot of code in uutils written by people who weren't aware of that. The coreutils generally do not require valif utf-8, which means that we cannot ever assume valid utf-8 and do an unchecked conversion, because that is undefined behaviour. So, those crates are indeed wrong.

Reply:
I found now a solution that works with crate `os_str_bytes` that significantly extends the functionality for OsStr. This way, one can use OsStr as a basis for parsing.